haystack/generator/seq2seq.py

```diff
@@ -0,0 +1,204 @@
+        :param use_gpu: Whether to use GPU (if available)
+        """
+
+        self.model_name_or_path = model_name_or_path
+        self.max_length = max_length
+        self.min_length = min_length
+        self.num_beams = num_beams
+        self.retriever = retriever
+
+        if top_k > self.num_beams:
+            top_k = self.num_beams
+            logger.warning(f'top_k value should not be greater than num_beams, hence setting it to {num_beams}')
+
+        self.top_k = top_k
+
+        if use_gpu and torch.cuda.is_available():
+            self.device = torch.device("cuda")
+        else:
+            self.device = torch.device("cpu")
+
+        self.tokenizer = AutoTokenizer.from_pretrained(model_name_or_path)
+        self.model = AutoModelForSeq2SeqLM.from_pretrained(model_name_or_path).to(self.device)
+        self.model.eval()
+
+    @abstractmethod
+    def prepare_model_input(self, query: str, documents: List[Document], top_k: Optional[int] = None) -> Dict:
+        pass
+
+    def predict(self, query: str, documents: List[Document], top_k: Optional[int] = None) -> Dict:
+        """
+        Generate the answer to the input query. The generation will be conditioned on the supplied documents.
+        These document can be retrieved via the Retriever or supplied directly via predict method.
+
+        :param query: Query
+        :param documents: Related documents (e.g. coming from a retriever) that the answer shall be conditioned on.
+        :param top_k: Number of returned answers
+        :return: Generated answers
+
+        """
+        torch.set_grad_enabled(False)
+        if len(documents) == 0:
+            raise AttributeError("generator needs documents to predict the answer")
+
+        top_k = top_k if top_k is not None else self.top_k
+
+        if top_k > self.num_beams:
+            top_k = self.num_beams
+            logger.warning(f'top_k value should not be greater than num_beams, hence setting it to {top_k}')
+
+        query_and_docs_encoded = self.prepare_model_input(query, documents, top_k)
+
+        generated_answers_encoded = self.model.generate(
+            input_ids=query_and_docs_encoded["input_ids"],
+            attention_mask=query_and_docs_encoded["attention_mask"],
+            min_length=self.min_length,
+            max_length=self.max_length,
+            do_sample=True if self.num_beams == 1 else False,
+            early_stopping=True,
+            num_beams=self.num_beams,
+            temperature=1.0,
+            top_k=None,
+            top_p=None,
+            eos_token_id=self.tokenizer.eos_token_id,
+            no_repeat_ngram_size=3,
+            num_return_sequences=top_k,
+            decoder_start_token_id=self.tokenizer.bos_token_id
+        )
+        generated_answers = self.tokenizer.batch_decode(generated_answers_encoded, skip_special_tokens=True)
+        return {"query": query, "answers": generated_answers}
+
+
+class BartEli5Generator(Seq2SeqGenerator):
```

Comment:
Not sure it is good idea to create separate class for `prepare_model_input` function.
I don't have much knowledge about model family though. @tholor your comment please